lib/src/main/java/macrobase/datamodel/DataFrame.java

```diff
@@ -0,0 +1,311 @@
+        if (numRows == 0) {
+            numRows = colValues.length;
+        }
+        schema.addColumn(ColType.DOUBLE, colName);
+        addDoubleColumnInternal(colValues);
+        return this;
+    }
+    private void addStringColumnInternal(String[] colValues) {
+        stringCols.add(colValues);
+        indexToTypeIndex.add(stringCols.size()-1);
+    }
+    public DataFrame addStringColumn(String colName, String[] colValues) {
+        if (numRows == 0) {
+            numRows = colValues.length;
+        }
+        schema.addColumn(ColType.STRING, colName);
+        addStringColumnInternal(colValues);
+        return this;
+    }
+
+    protected int[] getSubIndices(List<Integer> columns) {
+        int d = columns.size();
+        int[] typeSubIndices = new int[d];
+        for (int i = 0; i < d; i++) {
+            typeSubIndices[i] = indexToTypeIndex.get(columns.get(i));
+        }
+        return typeSubIndices;
+    }
+
+    public double[] getDoubleColumn(int columnIdx) {
+        return doubleCols.get(indexToTypeIndex.get(columnIdx));
+    }
+    public double[] getDoubleColumnByName(String columnName) {
+        return doubleCols.get(indexToTypeIndex.get(schema.getColumnIndex(columnName)));
+    }
+    public ArrayList<double[]> getDoubleCols(List<Integer> columns) {
+        ArrayList<double[]> cols = new ArrayList<>();
+        for (int c : columns) {
+            cols.add(getDoubleColumn(c));
+        }
+        return cols;
+    }
+    public ArrayList<double[]> getDoubleColsByName(List<String> columns) {
+        return getDoubleCols(this.schema.getColumnIndices(columns));
+    }
+    public String[] getStringColumn(int columnIdx) {
+        return stringCols.get(indexToTypeIndex.get(columnIdx));
+    }
+    public String[] getStringColumnByName(String columnName) {
+        return stringCols.get(indexToTypeIndex.get(schema.getColumnIndex(columnName)));
+    }
+    public ArrayList<String[]> getStringCols(List<Integer> columns) {
+        ArrayList<String[]> cols = new ArrayList<>();
+        for (int c : columns) {
+            cols.add(getStringColumn(c));
+        }
+        return cols;
+    }
+    public ArrayList<String[]> getStringColsByName(List<String> columns) {
+        return getStringCols(this.schema.getColumnIndices(columns));
+    }
+
+    /**
+     * @param columns column indices to project
+     * @return new dataframe with subset of columns
+     */
+    public DataFrame select(List<Integer> columns) {
+        DataFrame other = new DataFrame();
+        for (int c : columns) {
+            String columnName = schema.getColumnName(c);
+            ColType t = schema.getColumnType(c);
+            if (t == ColType.STRING) {
+                other.addStringColumn(columnName, getStringColumn(c));
+            } else if (t == ColType.DOUBLE) {
+                other.addDoubleColumn(columnName, getDoubleColumn(c));
+            } else {
+                throw new RuntimeException("Bad Column Type");
```

Comment:
Eep ugly! Throw `MacroBaseException` or similar?

RuntimeException is terrible for prod code